### ISSUE TITLE
Remove separate `CreateWebhookSource` AST node

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#20574]
*Part of #19951.*

The initial webhook implementation (https://github.com/MaterializeInc/materialize/pull/20171) added support for parsing `CREATE SOURCE ... FROM WEBHOOK` using a separate parsing path/AST node called `CreateWebhookSource`.

This doesn't seem right to me. Certainly the parsing of `CREATE SOURCE ... FROM WEBHOOK` should follow the same path as the existing sources—and ideally as much of the planning as possible too.

This ties into @parkerhendo's concern about webhook sources not matching users' existing mental model of what a source is: https://materializeinc.slack.com/archives/C02FWJ94HME/p1689469336240849

**Comment by benesch** [comment]
Ah, looks like there are a bunch of TODOs about merging the code paths once we put webhook sources on `clusterd`s. I think, for backwards compatibility, we should _pretend_ that webhook sources already run on `clusterd`s, require that you specify a cluster with `IN CLUSTER`, and then just ... not do that, for the initial implementation.

**Comment by ParkMyCar** [comment]
> looks like there are a bunch of TODOs about merging the code paths once we put webhook sources on clusterds.

Yeah since we don't require a cluster today I didn't want to make the [cluster_config](https://github.com/MaterializeInc/materialize/blob/8bd302f369a6a902679e34d68398ebb45a0c506a/src/sql/src/plan.rs#L528) optional/have a None variant.

> we should pretend that webhook sources already run on clusterds, require that you specify a cluster with IN CLUSTER, and then just ... not do that, for the initial implementation.

This makes sense to me, echoing what Cara said though, should we make the cluster they specify a source cluster? Or just totally ignore the cluster altogether?

> I say backwards incompat b/c it will suddenly start using compute resources on whichever cluster was specified in IN CLUSTER, which may OOM it or for some other reason not work.

I could be wrong, but I don't actually think this will be too much of a problem. The original "v1" webhooks will have minimal computation to do, at most they'll have to do request validation. It's once we introduce `TRANSFORM USING` that there could be non-trivial computation, but that will only be for newly created webhook sources. So I think we should be safe from randomly OOMing clusters. Maybe as a safeguard we could impose stricter request limits based on the size of the cluster? e.g. if they're creating a webhook source "on" a 3xsmall, then we limit it to 20 QPS?

**Comment by benesch** [comment]
> should we make the cluster they specify a source cluster

Ah, I'd say definitely make it a source cluster, yeah! We want this to look as much as possible like other sources.

> > I say backwards incompat b/c it will suddenly start using compute resources on whichever cluster was specified in IN CLUSTER, which may OOM it or for some other reason not work.
> 
> I could be wrong, but I don't actually think this will be too much of a problem. The original "v1" webhooks will have minimal computation to do, at most they'll have to do request validation. It's once we introduce `TRANSFORM USING` that there could be non-trivial computation, but that will only be for newly created webhook sources. So I think we should be safe from randomly OOMing clusters.

Agree with @ParkMyCar! I think the webhook resource requirements are small enough that this won't be a problem in practice. Potentially a good idea to leave webhook sources in "private preview" or "public preview" until we can actually run them on clusters, though. The thing that will really move the needle on resource requirements is supporting `ENVELOPE UPSERT`, but that _requires_ running them on clusters, and so we're forced to do things in the backwards-compatible order there.

**Comment by chaas** [comment]
Sounds good! Thanks for confirming.
This feels similar to how we "require" a cluster for `SELECT 1` queries (that we now "auto-route" to the `mz_introspection` cluster) even though the computation all happens in environmentd.

**Comment by benesch** [comment]
> This feels similar to how we "require" a cluster for `SELECT 1` queries (that we now "auto-route" to the `mz_introspection` cluster) even though the computation all happens in environmentd.

Agreed!

**Comment by benesch** [comment]
> I believe that covers the original concerns of this ticket, so I'm going to close it, but we can re-open if need be!

Ah my concern _is_ specifically about the parsing! Glad the planning got sorted out, but something still seems wrong about our AST nodes for `CREATE SOURCE` if we need to special case webhooks but not Kafka or Postgres sources. By my assessment, webhooks aren't any special-er than Kafka or Postgres sources! E.g., only Kafka has keys.

In my ideal world, we'd redesign the `CREATE SOURCE` AST like so:

```struct
enum Statement {
    CreateSource(CreateSourceStatement)
}

struct CreateSourceStatement {
    if_not_exists: bool,
    name: ObjectName,
    cluster: Option<Ident>
    from: CreateSourceFrom
}

enum CreateSourceFrom {
    Webhook(CreateSourceFromWebhook),
    Kafka(CreateSourceFromKafka),
    Postgres(CreateSourceFromPostgres),
}
```

No particular need for this to be a blocking item in the webhook epic, but @ParkMyCar if you don't mind I'd like to keep it open to track the long term cleanup of the tech debt.
